File: Other Versions/Go/scanners/aider.go
package scanners

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

var aiderSecretKeyTokens = []string{"key", "token", "secret", "password", "passwd", "auth", "credential"}

type aiderScanner struct{}

func init() {
	Register(&aiderScanner{})
}

func (s *aiderScanner) Name() string        { return "Aider" }
func (s *aiderScanner) Slug() string         { return "aider" }
func (s *aiderScanner) IsApplicable() bool   { return true }

func (s *aiderScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, p := range s.getConfigPaths(plat) {
		s.scanConfig(p, &result, showSecrets)
	}

	return result
}

func (s *aiderScanner) getConfigPaths(plat core.Platform) []string {
	var paths []string
	home := core.GetHome()
	paths = append(paths, filepath.Join(home, ".aider.conf.yml"))
	paths = append(paths, filepath.Join(home, ".aider.conf.yaml"))

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".aider.conf.yml"))
			paths = append(paths, filepath.Join(winHome, ".aider.conf.yaml"))
		}
	}
	return paths
}

func (s *aiderScanner) scanConfig(path string, result *core.ScanResult, showSecrets bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	storage := core.PlaintextYAML

	for _, rawLine := range strings.Split(string(data), "\n") {
		stripped := strings.TrimSpace(rawLine)
		if stripped == "" || strings.HasPrefix(stripped, "#") {
			continue
		}
		if !strings.Contains(stripped, ":") {
			continue
		}
		idx := strings.Index(stripped, ":")
		key := strings.TrimSpace(stripped[:idx])
		value := strings.TrimSpace(stripped[idx+1:])
		if value == "" {
			continue
		}

		// Strip inline comments (if value isn't quoted)
		if strings.Contains(value, "#") {
			if !(strings.HasPrefix(value, "\"") || strings.HasPrefix(value, "'")) {
				value = strings.TrimSpace(strings.SplitN(value, "#", 2)[0])
			}
		}

		// Strip surrounding quotes
		if len(value) >= 2 {
			first := value[0]
			last := value[len(value)-1]
			if first == last && (first == '"' || first == '\'') {
				value = value[1 : len(value)-1]
			}
		}
		if value == "" {
			continue
		}

		keyLower := strings.ToLower(key)
		matched := false
		for _, tok := range aiderSecretKeyTokens {
			if strings.Contains(keyLower, tok) {
				matched = true
				break
			}
		}
		if !matched {
			continue
		}

		valueLower := strings.ToLower(value)
		if valueLower == "true" || valueLower == "false" || valueLower == "yes" || valueLower == "no" || valueLower == "null" || valueLower == "~" {
			continue
		}

		notes := []string{fmt.Sprintf("Aider config key: %s", key)}
		if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
			notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
		}

		rawValue := ""
		if showSecrets {
			rawValue = value
		}

		result.Findings = append(result.Findings, core.CredentialFinding{
			ToolName:        s.Name(),
			CredentialType:  fmt.Sprintf("aider:%s", key),
			StorageType:     storage,
			Location:        path,
			Exists:          true,
			RiskLevel:       core.AssessRisk(storage, path),
			ValuePreview:    core.MaskValue(value, showSecrets),
			RawValue:        rawValue,
			FilePermissions: perms,
			FileOwner:       owner,
			FileModified:    core.GetFileMtime(path),
			Remediation:     "Use environment variables (OPENAI_API_KEY, ANTHROPIC_API_KEY) instead of config file",
			RemediationHint: remediation.HintMigrateToEnv([]string{"OPENAI_API_KEY", "ANTHROPIC_API_KEY"}, path),
			Notes:           notes,
		})
	}
}

File: aihound/scanners/aider.py
"""Scanner for Aider configuration files."""

from __future__ import annotations

import logging
from pathlib import Path

logger = logging.getLogger("aihound.scanners.aider")

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType,
)
from aihound.core.platform import detect_platform, Platform, get_home, get_wsl_windows_home
from aihound.core.redactor import mask_value
from aihound.core.permissions import (
    get_file_permissions, get_file_owner, assess_risk,
    get_file_mtime, describe_staleness,
)
from aihound.remediation import hint_migrate_to_env
from aihound.scanners import register


SECRET_KEY_TOKENS = ("key", "token", "secret", "password", "passwd", "auth", "credential")


@register
class AiderScanner(BaseScanner):
    def name(self) -> str:
        return "Aider"

    def slug(self) -> str:
        return "aider"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        for path in self._get_config_paths(plat):
            self._scan_config(path, result, show_secrets)

        return result

    def _get_config_paths(self, plat: Platform) -> list[Path]:
        paths: list[Path] = []
        home = get_home()

        # Linux/macOS/Windows: ~/.aider.conf.yml (and .yaml)
        paths.append(home / ".aider.conf.yml")
        paths.append(home / ".aider.conf.yaml")

        # WSL: also check Windows user home
        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".aider.conf.yml")
                paths.append(win_home / ".aider.conf.yaml")

        return paths

    def _scan_config(
        self, path: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not path.exists():
            logger.debug("Aider config not found: %s", path)
            return

        logger.debug("Reading Aider config: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)
        storage = StorageType.PLAINTEXT_YAML

        try:
            content = path.read_text(encoding="utf-8")
        except OSError as e:
            logger.warning("Failed to read %s: %s", path, e)
            result.errors.append(f"Failed to read {path}: {e}")
            return

        try:
            for line in content.splitlines():
                stripped = line.strip()
                if not stripped or stripped.startswith("#"):
                    continue
                if ":" not in stripped:
                    continue

                key, _, value = stripped.partition(":")
                key = key.strip()
                value = value.strip()

                if not value:
                    continue

                # Strip inline comments
                if "#" in value:
                    # Only strip if not inside quotes; simple heuristic
                    if not (value.startswith('"') or value.startswith("'")):
                        value = value.split("#", 1)[0].strip()

                # Strip surrounding quotes
                if len(value) >= 2 and value[0] == value[-1] and value[0] in ('"', "'"):
                    value = value[1:-1]

                if not value:
                    continue

                key_lower = key.lower()
                if not any(tok in key_lower for tok in SECRET_KEY_TOKENS):
                    continue

                # Skip obvious non-secret values: booleans, numbers
                if value.lower() in ("true", "false", "yes", "no", "null", "~"):
                    continue

                notes = [f"Aider config key: {key}"]
                if mtime:
                    notes.append(f"File last modified: {describe_staleness(mtime)}")

                result.findings.append(CredentialFinding(
                    tool_name=self.name(),
                    credential_type=f"aider:{key}",
                    storage_type=storage,
                    location=str(path),
                    exists=True,
                    risk_level=assess_risk(storage, path),
                    value_preview=mask_value(value, show_full=show_secrets),
                    raw_value=value if show_secrets else None,
                    file_permissions=perms,
                    file_owner=owner,
                    file_modified=mtime,
                    remediation="Use environment variables (OPENAI_API_KEY, ANTHROPIC_API_KEY) instead of config file",
                    remediation_hint=hint_migrate_to_env(["OPENAI_API_KEY", "ANTHROPIC_API_KEY"], path),
                    notes=notes,
                ))
        except Exception as e:
            logger.warning("Failed to parse %s: %s", path, e, exc_info=True)
            result.errors.append(f"Failed to parse {path}: {e}")
